File: gateway-demo/src/main/java/reactivekitchen/bank/gateway/GatewayVerticle.java
package reactivekitchen.bank.gateway;

import io.reactivex.Single;
import io.vertx.core.json.JsonObject;
import io.vertx.reactivex.core.AbstractVerticle;
import io.vertx.reactivex.ext.web.Router;
import io.vertx.reactivex.ext.web.RoutingContext;
import io.vertx.reactivex.ext.web.client.HttpResponse;
import io.vertx.reactivex.ext.web.client.WebClient;
import reactivekitchen.bank.gateway.helpers.Services;


public class GatewayVerticle extends AbstractVerticle {

  private WebClient service;
  private WebClient debt;
  private WebClient balance;

  @Override
  public void start() throws Exception {

    service = Services.getCustomerService(vertx);
    debt = Services.getDebtService(vertx);
    balance = Services.getBalanceService(vertx);

    Router router = Router.router(vertx);
    router.get("/").handler(this::compute);
    // TODO GET / -> compute

    vertx.createHttpServer().requestHandler(router::accept).listen(8080);
  }

  private void compute(RoutingContext rc) {

    Single<JsonObject> s = service.get("/customers/clement")
        .rxSend()
        .map(HttpResponse::bodyAsJsonObject).flatMap(json -> {
          String account = json.getString("account");

          Single<Double> s1 = balance.get("/balance/" + account)
              .rxSend()
              .map(HttpResponse::bodyAsJsonObject)
              .map(j -> j.getDouble("balance"));

          Single<Double> s2 = debt.get("/debt/" + account)
              .rxSend()
              .map(HttpResponse::bodyAsJsonObject)
              .map(j -> j.getDouble("level"));


          return Single.zip(s1, s2, (balance, debt) -> new JsonObject()
              .put("account", account)
              .put("balance", balance)
              .put("debt", debt));
        });
    s.subscribe(json -> {
      rc.response().end(json.encodePrettily());
    }, err -> {
      err.printStackTrace();
      rc.response().end("ERROR " + err.getMessage());
    });

    // TODO Get customer account ("/customers/clement") -> account
    // TODO Get balance /balance/$account -> balance
    // TODO Get debt /debt/$account -> level


  }
}

File: gateway-demo/src/main/kotlin/com/julienviet/demo/bank/gateway/GatewayVerticle.kt
package com.julienviet.demo.bank.gateway

import io.vertx.core.AbstractVerticle
import io.vertx.core.buffer.Buffer
import io.vertx.ext.web.Router
import io.vertx.ext.web.client.HttpResponse
import io.vertx.ext.web.client.WebClient
import io.vertx.kotlin.core.json.json
import io.vertx.kotlin.core.json.obj
import io.vertx.kotlin.coroutines.awaitResult
import io.vertx.kotlin.coroutines.dispatcher
import kotlinx.coroutines.experimental.launch
import me.escoffier.demo.bank.gateway.helpers.Services

class GatewayVerticle : AbstractVerticle() {

  lateinit var service: WebClient
  lateinit var debt: WebClient
  lateinit var balance: WebClient

  override fun start() {

    service = Services.getCustomerService(vertx)
    debt = Services.getDebtService(vertx)
    balance = Services.getBalanceService(vertx)

    val router = Router.router(vertx)
    // TODO GET / -> compute
    router.get("/").handler({
      launch(vertx.dispatcher()) {
        try {
          compute(it)
        } catch (e: Exception) {
          e.printStackTrace()
        }
      }
    })

    vertx.createHttpServer()
        .requestHandler { router.accept(it) }
        .listen(8080)
  }

  private suspend fun compute(rc: io.vertx.ext.web.RoutingContext) {

    val accountResp = awaitResult<HttpResponse<Buffer>> {
      service
          .get("/customers/clement")
          .send(it)
    }

    val account = accountResp.bodyAsJsonObject().getString("account")

    val debtResp = awaitResult<HttpResponse<Buffer>> { debt.get("/debt/" + account).send(it) }
    val balanceResp = awaitResult<HttpResponse<Buffer>> { balance.get("/balance/" + account).send(it) }

    val debt = debtResp.bodyAsJsonObject().getDouble("level")
    val balance = balanceResp.bodyAsJsonObject().getDouble("balance")
    val json = json {
      obj(
          "account" to account,
          "debt" to debt,
          "balance" to balance
      )
    }

    rc.response().end(json.encodePrettily())
  }
}